GVRf/Framework/framework/src/main/java/org/gearvrf/GVRJassimpAdapter.java

```diff
@@ -408,93 +409,388 @@ public void processBones(GVRMesh mesh, List<AiBone> aiBones)
+            channel.setPosKeyVector(0, t, curpos);
+            for (i = 1; i < aiNodeAnim.getNumPosKeys(); ++i)
+            {
+                float[] pos = aiNodeAnim.getPosKeyVector(i, sWrapperProvider);
+                if (!isEqual(pos, curpos))
+                {
+                    t = (float) aiNodeAnim.getPosKeyTime(i) / ticksPerSec;
+                    channel.setPosKeyVector(nextIndex++, t, pos);
+                    curpos = pos;
+                }
+            }
+            channel.resizePosKeys(nextIndex);
         }
 
-        // Rot keys
-        for (i = 0; i < aiNodeAnim.getNumRotKeys(); ++i) {
-            q = aiNodeAnim.getRotKeyQuaternion(i, sWrapperProvider);
-            float[] rot = new float[] { q.x, q.y, q.z, q.w };
-            node.setRotKeyQuaternion(i, (float)aiNodeAnim.getRotKeyTime(i), rot);
+        if (aiNodeAnim.getNumRotKeys() > 0)
+        {
+            Quaternionf currot = aiNodeAnim.getRotKeyQuaternion(0, sWrapperProvider);
+            int nextIndex = 1;
+
+            t = (float) aiNodeAnim.getRotKeyTime(0) / ticksPerSec;
+            channel.setRotKeyQuaternion(0, t, currot);
+            for (i = 1; i < aiNodeAnim.getNumRotKeys(); ++i)
+            {
+                Quaternionf rot = aiNodeAnim.getRotKeyQuaternion(i, sWrapperProvider);
+                if (!isEqual(rot, currot))
+                {
+                    t = (float) aiNodeAnim.getRotKeyTime(i) / ticksPerSec;
+                    channel.setRotKeyQuaternion(nextIndex++, t, rot);
+                    currot = rot;
+                }
+            }
+            channel.resizeRotKeys(nextIndex);
         }
 
-        // Scale keys
-        for (i = 0; i < aiNodeAnim.getNumScaleKeys(); ++i) {
-            float[] scale = aiNodeAnim.getScaleKeyVector(i, sWrapperProvider);
-            node.setScaleKeyVector(i, (float)aiNodeAnim.getScaleKeyTime(i), scale);
+        if (aiNodeAnim.getNumScaleKeys() > 0)
+        {
+            int nextIndex = 1;
+            float[] curscale = aiNodeAnim.getScaleKeyVector(0, sWrapperProvider);
+
+            t = (float) aiNodeAnim.getScaleKeyTime(0) / ticksPerSec;
+            channel.setScaleKeyVector(0, t, curscale);
+            for (i = 1; i < aiNodeAnim.getNumScaleKeys(); ++i)
+            {
+                float[] scale = aiNodeAnim.getScaleKeyVector(i, sWrapperProvider);
+
+                if (!isEqual(scale, curscale))
+                {
+                    t = (float) aiNodeAnim.getScaleKeyTime(i) / ticksPerSec;
+                    channel.setScaleKeyVector(nextIndex++, t, scale);
+                }
+            }
+            channel.resizeScaleKeys(nextIndex);
         }
+        return channel;
+    }
 
-        return node;
+    private boolean isEqual(float[] arr1, float[] arr2)
+    {
+        final float EPSILON = 0.0001f;
+
+        for (int i = 0; i < arr1.length; ++i)
+        {
+            if (Math.abs(arr2[i] - arr1[i]) > EPSILON)
```

Comment:
There is also Float.compare..

Reply:
If it is not an exact compare I will use it

Reply:
I see. Then these epsilon-based methods need to deal with the edge cases raised by NaNs and the like.